ironfish/src/primitives/blockheader.ts

```diff
@@ -237,62 +239,48 @@ export class BlockHeaderSerde implements Serde<BlockHeader, SerializedBlockHeade
     )
   }
 
-  serialize(header: BlockHeader): SerializedBlockHeader {
-    const serialized = {
-      sequence: header.sequence,
-      previousBlockHash: BlockHashSerdeInstance.serialize(header.previousBlockHash),
-      noteCommitment: {
-        commitment: this.strategy.noteHasher
-          .hashSerde()
-          .serialize(header.noteCommitment.commitment),
-        size: header.noteCommitment.size,
-      },
-      nullifierCommitment: {
-        commitment: this.strategy.nullifierHasher
-          .hashSerde()
-          .serialize(header.nullifierCommitment.commitment),
-        size: header.nullifierCommitment.size,
-      },
-      target: header.target.targetValue.toString(),
-      randomness: header.randomness,
-      timestamp: header.timestamp.getTime(),
-      minersFee: header.minersFee.toString(),
-      work: header.work.toString(),
-      hash: BlockHashSerdeInstance.serialize(header.hash),
-      graffiti: GraffitiSerdeInstance.serialize(header.graffiti),
-    }
-
-    return serialized
+  serialize(header: BlockHeader): Buffer {
+    const bw = bufio.write()
+    bw.writeU64(header.sequence)
+    bw.writeHash(header.previousBlockHash)
+    bw.writeHash(header.noteCommitment.commitment)
+    bw.writeU64(header.noteCommitment.size)
+    bw.writeHash(header.nullifierCommitment.commitment)
+    bw.writeU64(header.nullifierCommitment.size)
+    bw.writeBytes(BigIntUtils.toBytesBE(header.target.asBigInt(), 32))
+    bw.writeU64(header.randomness)
+    bw.writeU64(header.timestamp.getTime())
+    bw.writeBytes(BigIntUtils.toBytesBE(header.minersFee, 8))
+    bw.writeVarBytes(header.graffiti)
+    bw.writeVarBytes(BigIntUtils.toBytes(header.work))
+    bw.writeHash(header.hash)
+    return bw.render()
   }
 
-  deserialize(data: SerializedBlockHeader): BlockHeader {
+  deserialize(data: Buffer): BlockHeader {
     // TODO: this needs to make assertions on the data format
     // as it can be from untrusted sources
+    const br = bufio.read(data)
     const header = new BlockHeader(
       this.strategy,
-      Number(data.sequence),
-      Buffer.from(BlockHashSerdeInstance.deserialize(data.previousBlockHash)),
+      Number(br.readU64()),
+      br.readHash(),
       {
-        commitment: this.strategy.noteHasher
-          .hashSerde()
-          .deserialize(data.noteCommitment.commitment),
-        size: data.noteCommitment.size,
+        commitment: br.readHash(),
```

Comment:
Reading in an object literal like this can be dangerous because the keys could get read in the incorrect order. I think it's better to read all the data out then construct this object.